[libc++] use copy_file_range for fs::copy

Jannik2099 opened the pull request:
Hi,

this optimizes `std::filesystem::copy_file` to use the `copy_file_range` syscall (Linux and FreeBSD) when available. It allows for reflinks on filesystems such as btrfs, zfs and xfs, and server-side copy for network filesystems such as NFS.

This behaviour is in line with Boost.Filesystem and libstdc++ (where [I implemented this](https://github.com/gcc-mirror/gcc/commit/d87caacf8e2df563afda85f3a5b7b852e08b6b2c) some time ago). It's also the default behaviour of the standard file copy operations in .NET Core, Java, and (hopefully) soon python, and in the `cp` program from GNU coreutils.

I don't have the opportunity to test this on FreeBSD, but I think the FreeBSD folks will want a look at this. Sadly I don't know any by name, perhaps someone could CC the relevant person(s).

As for the `LIBCXX_USE_COPY_FILE_RANGE` cmake check, I was unsure where to put it, and this is probably not the desired location.

Lastly I am not sure if the largefile handling (see usage of `off_t`) is correct.

Jannik2099 committed:
[libc++] use copy_file_range for fs::copy_file

opportunistically use copy_file_range (Linux, FreeBSD) where possible.
This allows for fast copies via reflinks,
and server side copies for network filesystems.
Fall back to sendfile if not supported.

Signed-off-by: Jannik Glückert <jannik.glueckert@gmail.com>

mstorsjo commented:
> I don't have the opportunity to test this on FreeBSD, but I think the FreeBSD folks will want a look at this. Sadly I don't know any by name, perhaps someone could CC the relevant person(s).

CC @emaste 

@Jannik2099 Note that the CI environment does contain a FreeBSD setup, which shows a failure (a somewhat boring one):
```
/home/buildkite/.buildkite-agent/builds/freebsd-test-1/llvm-project/libcxx-ci/libcxx/src/filesystem/operations.cpp:183:6: error: unused function 'copy_file_impl_copy_file_range' [-Werror,-Wunused-function]
  183 | bool copy_file_impl_copy_file_range(FileDescriptor& read_fd, FileDescriptor& write_fd, error_code& ec) {
      |      ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
1 error generated.
```
(There are also a couple other similar failing cases in the same CI run on buildkite.)

Jannik2099 commented:
Thanks!

Yeah, I forgot to slap on `-Werror` yesterday, will fix later.

Another note, how TOCTOU-safe should this be? Right now I fstat both in the copy_file_range and in the sendfile impl. If the file size changes inbetween and copy_file_range fails halfway through, this may lead to inconsistent output.

Jannik2099 commented:
So I looked at the FreeBSD failure again...

FreeBSD does have `sendfile`, but it does NOT have `sys/sendfile.h`, which means libc++ has been doing userspace copies on FreeBSD all this time!

I'll think of a more robust check.

Jannik2099 commented:
> FreeBSD does have sendfile, but it does NOT have sys/sendfile.h, which means libc++ has been doing userspace copies on FreeBSD all this time!

so turns out that was on purpose, because FreeBSD `sendfile` cannot send to files - I had assumed it was one of those "ancient unix syscalls that made it's way into every OS".

I'll have to restructure this a bit then.

Unrelated, I also took the opportunity to support copying special zero-length files on linux (such as those in `/proc`). Funnily enough, we discovered the same deficiency in libstdc++ when I was adding `copy_file_range` support there.

Jannik2099 committed:
[libc++] support special linux files in fs::copy_file

Virtual linux filesystems such as /proc and /sys
may contain files that appear as zero length,
but do contain data.
These require a traditional userspace read + write loop.

Signed-off-by: Jannik Glückert <jannik.glueckert@gmail.com>

Jannik2099 commented:
Ok, I think this is ready for review now.

To summarize:

Add `copy_file_range` support on Linux and FreeBSD. `fs::copy_file` will first try `copy_file_range`, if unavailable and on linux `sendfile`, and finally the generic fstream.

This covers the following scenarios:

* `copy_file_range` is not implemented by the filesystem or the OS
* `copy_file_range` and/or `sendfile` are unable to copy to/from the source/dest in question
* `fs::copy_file` is used on a special linux zero-length file, such as those in `/proc` - this was previously unsupported.

I have also added a test for the `/proc` case, there are no such files on FreeBSD to my rudimentary knowledge, so the test is linux exclusive.

I am not super happy with how `copy_file_impl` looks now - it's stuck halfway between doing `#ifdef` on (effectively) the target OS, and halfway between doing it on the available syscalls. Feel free to tear apart.

ldionne commented:
Please rebase onto main to re-trigger CI. The CI instability should be resolved now.

ldionne commented:
(I'll get to this review in my queue, for now just trying to get CI back on track)

ldionne commented on libcxx/test/std/input.output/filesystems/fs.op.funcs/fs.op.copy_file/copy_file_procfs.pass.cpp:
```diff
@@ -0,0 +1,50 @@
+#include "test_macros.h"
+#include "filesystem_test_helper.h"
+
+using namespace std::filesystem;
```
Please use `namespace fs = std::filesystem` instead. We generally discourage `using namespace foo;` (I know some places in the test suite don't follow that).

ldionne commented on libcxx/test/std/input.output/filesystems/fs.op.funcs/fs.op.copy_file/copy_file_procfs.pass.cpp:
```diff
@@ -0,0 +1,50 @@
+// This is because the to-be-read data is usually generated ad-hoc by the reading syscall
+// These files can not be copied with kernel-side copies like copy_file_range or sendfile,
+// and must instead be copied via a traditional userspace read + write loop.
+int main(int, char**) {
```
Nitpick: `main` must always explicitly `return 0;` in our tests, because that otherwise breaks tests under `-ffreestanding` (since under that flag `main` is not treated as a special function).

ldionne commented on libcxx/test/std/input.output/filesystems/fs.op.funcs/fs.op.copy_file/copy_file_procfs.pass.cpp:
```diff
@@ -0,0 +1,50 @@
+// <filesystem>
+
+// bool copy_file(const path& from, const path& to);
+// bool copy_file(const path& from, const path& to, error_code& ec) noexcept;
```
Right now, you're only testing this overload of the function. Could you update the test to test the other variants as well?

Jannik2099 replied:
The `copy_file.pass.cpp` test already covers these - `copy_file_large.pass.cpp` skips them aswell

ldionne commented on libcxx/test/std/input.output/filesystems/fs.op.funcs/fs.op.copy_file/copy_file_procfs.pass.cpp:
```diff
@@ -0,0 +1,50 @@
+//           error_code& ec) noexcept;
+
+#include <filesystem>
+#include <cassert>
```
Missing `#include <system_error>`.

ldionne commented on libcxx/test/std/input.output/filesystems/fs.op.funcs/fs.op.copy_file/copy_file_procfs.pass.cpp:
```diff
@@ -0,0 +1,50 @@
+  assert(copy_file(procfile, dest, ec));
+  assert(!ec);
+
+  // a tad fragile if lit ever changes the output name,
+  // but the easiest way to assert that the correct data was copied.
+  assert(file_size(dest) == sizeof("t.tmp.exe"));
```
Could you instead derive something from `argv[0]`?

ldionne commented on libcxx/src/filesystem/operations.cpp:
```diff
@@ -194,6 +271,46 @@ bool copy_file_impl(FileDescriptor& read_fd, FileDescriptor& write_fd, error_cod
+#endif
+
+#if defined(_LIBCPP_FILESYSTEM_USE_COPY_FILE_RANGE) || defined(_LIBCPP_FILESYSTEM_USE_SENDFILE)
+bool copy_file_impl(FileDescriptor& read_fd, FileDescriptor& write_fd, error_code& ec) {
+#  if defined(_LIBCPP_FILESYSTEM_USE_COPY_FILE_RANGE)
+  if (copy_file_impl_copy_file_range(read_fd, write_fd, ec)) {
+    return true;
+  }
+  // EINVAL: src and dst are the same file (this is not cheaply
+  // detectable from userspace)
+  // EINVAL: copy_file_range is unsupported for this file type by the
+  // underlying filesystem
+  // ENOTSUP: undocumented, can arise with old kernels and NFS
+  // EOPNOTSUPP: filesystem does not implement copy_file_range
+  // ETXTBSY: src or dst is an active swapfile (nonsensical, but allowed
+  // with normal copying)
+  // EXDEV: src and dst are on different filesystems that do not support
+  // cross-fs copy_file_range
+  // ENOENT: undocumented, can arise with CIFS
+  // ENOSYS: unsupported by kernel or blocked by seccomp
+  if (ec.value() != EINVAL && ec.value() != ENOTSUP && ec.value() != EOPNOTSUPP && ec.value() != ETXTBSY &&
+      ec.value() != EXDEV && ec.value() != ENOENT && ec.value() != ENOSYS) {
+    return false;
+  }
+  ec.clear();
+#  endif
+
+#  if defined(_LIBCPP_FILESYSTEM_USE_SENDFILE)
+  if (copy_file_impl_sendfile(read_fd, write_fd, ec)) {
+    return true;
+  }
+  // EINVAL: unsupported file type
+  if (ec.value() != EINVAL) {
+    return false;
+  }
+  ec.clear();
+#  endif
+
+  return copy_file_impl_fstream(read_fd, write_fd, ec);
+}
```
I would rewrite this as follows, I think that makes it easier to understand and we duplicate very little code:

```c++
#if defined(_LIBCPP_FILESYSTEM_USE_COPY_FILE_RANGE)

bool copy_file_impl(FileDescriptor& read_fd, FileDescriptor& write_fd, error_code& ec) {
  if (copy_file_impl_copy_file_range(read_fd, write_fd, ec)) {
    return true;
  }
  // EINVAL: src and dst are the same file (this is not cheaply
  // detectable from userspace)
  // EINVAL: copy_file_range is unsupported for this file type by the
  // underlying filesystem
  // ENOTSUP: undocumented, can arise with old kernels and NFS
  // EOPNOTSUPP: filesystem does not implement copy_file_range
  // ETXTBSY: src or dst is an active swapfile (nonsensical, but allowed
  // with normal copying)
  // EXDEV: src and dst are on different filesystems that do not support
  // cross-fs copy_file_range
  // ENOENT: undocumented, can arise with CIFS
  // ENOSYS: unsupported by kernel or blocked by seccomp
  if (ec.value() != EINVAL && ec.value() != ENOTSUP && ec.value() != EOPNOTSUPP && ec.value() != ETXTBSY &&
      ec.value() != EXDEV && ec.value() != ENOENT && ec.value() != ENOSYS) {
    return false;
  }
  ec.clear();

  return copy_file_impl_fstream(read_fd, write_fd, ec);
}

#elif defined(_LIBCPP_FILESYSTEM_USE_SENDFILE)

bool copy_file_impl(FileDescriptor& read_fd, FileDescriptor& write_fd, error_code& ec) {
  if (copy_file_impl_sendfile(read_fd, write_fd, ec)) {
    return true;
  }
  // EINVAL: unsupported file type
  if (ec.value() != EINVAL) {
    return false;
  }
  ec.clear();

  return copy_file_impl_fstream(read_fd, write_fd, ec);
}
 
#elif defined(_LIBCPP_FILESYSTEM_USE_COPYFILE)

...
```

And once you rewrite like that, you could also consider inlining `copy_file_impl_copy_file_range` and `copy_file_impl_sendfile` into their respective `copy_file_impl`. IDK if that would simplify the code but it might.

Jannik2099 replied:
That doesn't work, `copy_file_range` will (usually) fail on cross-fs copy, and we'd want to fall back to `sendfile` in that case, and not abandon ship and do a userspace copy.

Since only Linux and FreeBSD support `copy_file_range`, and only linux supports `sendfile` on files, perhaps something like:

```
#ifdef __linux__ || __FreeBSD__
copy_file_impl {
    #ifdef HAS_COPY_FILE_RANGE
    if(copy_file_range) { return; }
    #endif
    #ifdef HAS_SENDFILE
    if(sendfile) { return; }
    #endif
    fstream
}
#endif
```

Of course, this will get procedually more messy when another Unix platform implements any combination of `copy_file_range` and `sendfile`.

ldionne replied:
There is something I don't understand. `_LIBCPP_FILESYSTEM_USE_COPY_FILE_RANGE`, `_LIBCPP_FILESYSTEM_USE_SENDFILE` and `_LIBCPP_FILESYSTEM_USE_COPYFILE` are mutually exclusive (right?). My rewriting didn't change any logic at all, it only reorganized the code a bit to define the whole function inside each `#if` branch. Unless I missed something and my "refactoring" actually changed the logic?

Jannik2099 replied:
No, `_LIBCPP_FILESYSTEM_USE_COPY_FILE_RANGE` and `_LIBCPP_FILESYSTEM_USE_SENDFILE` are not mutually exclusive. The intent is to attempt `copy_file_range`, and fall back to `sendfile` otherwise.

ldionne replied:
Ah, I see what I missed now.

I would suggest something like this (which is basically what you wrote above IIUC):

```c++
#if HAS_SENDFILE
copy_file_impl_sendfile() { ... }
#endif

#if HAS_COPY_FILE_RANGE
copy_file_impl_copy_file_range() { ... }
#endif

#if HAS_COPY_FILE
copy_file_impl_copyfile() { ... }
#endif

copy_file_impl_fstream() { ... }


copy_file_impl() {
#if something
  full implementation
#elif something
  full implementation
#else
  full implementation
#endif
}
```

Does that make sense? Decouple the implementation of the different functions and the place where you compose them together, so that we can see at a glance what `copy_file_impl` does when looking at it.

Jannik2099 replied:
if you want the full implementation inlined into `copy_file_impl`, then what's the point of defining e.g. `copy_file_impl_sendfile` above?

ldionne replied:
By "full implementation", I simply meant that you wouldn't interleave code within the `#ifdef` and code outside the `#ifdef`s, I didn't mean that you would inline everything recursively into `copy_file_impl`.

Jannik2099 replied:
is this to your liking now? The `error_code` handling still resides in `copy_file_impl`, as otherwise the various impls would have to return some kind of tribool to disambiguate soft failure due to preconditions and hard failure due to another error

ldionne commented on libcxx/src/CMakeLists.txt:
```diff
@@ -173,6 +173,12 @@ if (APPLE AND LLVM_USE_SANITIZER)
   endif()
 endif()
 
+include(CheckCXXSymbolExists)
+check_cxx_symbol_exists("copy_file_range" "unistd.h" LIBCXX_USE_COPY_FILE_RANGE)
+if(LIBCXX_USE_COPY_FILE_RANGE)
+  list(APPEND LIBCXX_COMPILE_FLAGS "-D_LIBCPP_FILESYSTEM_USE_COPY_FILE_RANGE")
+endif()
```
What Linux versions don't provide `copy_file_range`? I'd rather assume support from the OS or test for something general like `#ifdef __linux__` in `operations.cpp` than add a CMake check for a single OS feature like this.

Jannik2099 replied:
since linux 4.5, according to the man page. RHEL 8 is at linux 4.18, for comparison. I don't think any still supported distro uses < 4.5
On FreeBSD, it has been added in FreeBSD 13, and FreeBSD 12.x went EoL on 31. December 2023

So yeah, if you want I'll just drop this and do an `#ifdef __linux__ || __FreeBSD__`

ldionne reviewed (changes requested):
Thanks for the patch and sorry for the delay! I have some comments and questions, but overall this patch looks quite good and this is definitely a nice optimization.

Jannik2099 commented:
@ldionne ping

I think all the remaining CI failures are unrelated, except for the one mentioned above. Though I don't know how to fix this.

ldionne committed:
Minor reorganization per review comments

ldionne commented on libcxx/test/std/input.output/filesystems/fs.op.funcs/fs.op.copy_file/copy_file_procfs.pass.cpp:
```diff
@@ -0,0 +1,52 @@
+//===----------------------------------------------------------------------===//
+
+// UNSUPPORTED: c++03, c++11, c++14
+// REQUIRES: linux
```
```suggestion
// REQUIRES: linux || freebsd
```

Since we use `copy_file_range` on FreeBSD, we should enable this test there.

Jannik2099 replied:
This test is for zero-length files in `/proc` specifically.

FreeBSD deprecated procfs a long while ago, and I don't know if it has any "virtual zero-length but not actually empty" files like linux does.

ldionne replied:
Ack, disregard my comment then.

ldionne reviewed (approved):
LGTM with minor comment. I've pushed to your branch to reorganize the way I wanted since it was only a minor deviation from what you did -- please LMK if what I did is not OK with you.

ldionne committed:
Make _LIBCPP_FILESYSTEM_USE_COPY_FILE_RANGE and _LIBCPP_FILESYSTEM_USE_SENDFILE disjoint

ldionne reviewed (approved):
LGTM once CI is green.

Jannik2099 commented:
> LGTM once CI is green.

NACK, see my review above

ldionne commented:
> > LGTM once CI is green.
> 
> NACK, see my review above

Sorry, I don't understand. Are you saying that something still needs to be addressed in this PR?

Jannik2099 commented:
> > > LGTM once CI is green.
> > 
> > 
> > NACK, see my review above
> 
> Sorry, I don't understand. Are you saying that something still needs to be addressed in this PR?

I am saying that your commit https://github.com/llvm/llvm-project/pull/109211/commits/8f314b977c7a5a986f68a97e9b29b7725770460d completely changes the semantics.

You changed it to two definitions of `copy_file_impl` that attempt `copy_file_range` / `sendfile` first, and `fstream` second. The definition is chosen at compile time.

The intended behaviour is for one `copy_file_impl` that attempts `copy_file_range` (if available for the platform), then `sendfile` (if available for the platform), and finally `fstream`

ldionne commented:
Oh, geez, now I understand the intent. That's subtle. I assumed the copy-file-range implementation and the sendfile implementations were disjoint.

ldionne committed:
Fix incorrect semantics

ldionne reviewed (approved):
I just pushed something that should fix my mistake.

Can you double-check and confirm that this does what you originally intended? I think we're pretty much back to your patch without my edits, except I added a comment.

If that patch is fine with you, it's fine with me in the current state. Sorry for the mixup, I was trying to help expedite this patch by doing what I thought were NFC changes (but were not).

Jannik2099 commented:
> Sorry for the mixup

No worries!

LGTM

Jannik2099 commented:
Ugh, I forgot that Bionic does not provide a `copy_file_range` syscall wrapper. I guess checking for availability of the function at configure time like I did initially is the best option?

Also, there's still the issue that the no-locale build now fails due to use of `fstream`. Does libc++ have an internal locale-independent `fstream` I could use here? It'd be silly to drop `fs::copy_file` support from the no-locale build, but so is duplicating the old implementation which doesn't support `/proc` files

ldionne commented:
> Ugh, I forgot that Bionic does not provide a `copy_file_range` syscall wrapper. I guess checking for availability of the function at configure time like I did initially is the best option?

Can't you just check for `#ifdef __BIONIC__`? I'd like to avoid introducing any CMake coupling just to check that property.

> 
> Also, there's still the issue that the no-locale build now fails due to use of `fstream`. Does libc++ have an internal locale-independent `fstream` I could use here? It'd be silly to drop `fs::copy_file` support from the no-locale build, but so is duplicating the old implementation which doesn't support `/proc` files

No, we don't have such a thing. Instead, I'd disable it for the no-localization build for now. I'm working on changes that should make `fstream` available even when localization is disabled, so that problem would be solved then.

Jannik2099 committed:
fix bionic and no-locale config

Signed-off-by: Jannik Glückert <jannik.glueckert@gmail.com>

Jannik2099 committed:
set appropriate error_code when failing due to lack of fstream

Jannik2099 committed:
fix no-locale empty file copy

ldionne commented:
Thanks for the patch!

vitalybuka commented:
We have this error now
https://lab.llvm.org/buildbot/#/builders/66/builds/8314

```
/home/b/sanitizer-x86_64-linux/build/build_default/bin/clang++ -DLIBCXX_BUILDING_LIBCXXABI -DLIBC_NAMESPACE=__llvm_libc_common_utils -D_LIBCPP_BUILDING_LIBRARY -D_LIBCPP_HAS_NO_PRAGMA_SYSTEM_HEADER -D_LIBCPP_LINK_PTHREAD_LIB -D_LIBCPP_LINK_RT_LIB -D_LIBCPP_REMOVE_TRANSITIVE_INCLUDES -D__STDC_CONSTANT_MACROS -D__STDC_FORMAT_MACROS -D__STDC_LIMIT_MACROS -I/home/b/sanitizer-x86_64-linux/build/llvm-project/libcxx/src -I/home/b/sanitizer-x86_64-linux/build/build_default/runtimes/runtimes-bins/compiler-rt/lib/sanitizer_common/symbolizer/RTSanitizerCommonSymbolizerInternal.i386/symbolizer/libcxx/include/c++/v1 -I/home/b/sanitizer-x86_64-linux/build/llvm-project/libcxxabi/include -I/home/b/sanitizer-x86_64-linux/build/llvm-project/runtimes/cmake/Modules/../../../libc -fvisibility-inlines-hidden -Werror=date-time -Werror=unguarded-availability-new -Wall -Wextra -Wno-unused-parameter -Wwrite-strings -Wcast-qual -Wmissing-field-initializers -Wimplicit-fallthrough -Wcovered-switch-default -Wno-noexcept-type -Wnon-virtual-dtor -Wdelete-non-virtual-dtor -Wsuggest-override -Wstring-conversion -Wmisleading-indentation -Wctad-maybe-unsupported -fdiagnostics-color -ffunction-sections -fdata-sections -m32 -U_FILE_OFFSET_BITS -fPIC -flto -Oz -g0 -DNDEBUG -target i386-unknown-linux-gnu -Wno-unused-command-line-argument -include /home/b/sanitizer-x86_64-linux/build/llvm-project/compiler-rt/lib/sanitizer_common/symbolizer/../sanitizer_redefine_builtins.h -DSANITIZER_COMMON_REDEFINE_BUILTINS_IN_STD -Wno-language-extension-token -Wno-macro-redefined -std=c++23 -faligned-allocation -nostdinc++ -fvisibility-inlines-hidden -fvisibility=hidden -fsized-deallocation -Wall -Wextra -Wnewline-eof -Wshadow -Wwrite-strings -Wno-unused-parameter -Wno-long-long -Werror=return-type -Wextra-semi -Wundef -Wunused-template -Wformat-nonliteral -Wzero-length-array -Wdeprecated-redundant-constexpr-static-def -Wno-user-defined-literals -Wno-covered-switch-default -Wno-suggest-override -Wno-error -fno-exceptions -fno-rtti -MD -MT libcxx/src/CMakeFiles/cxx_static.dir/filesystem/operations.cpp.o -MF libcxx/src/CMakeFiles/cxx_static.dir/filesystem/operations.cpp.o.d -o libcxx/src/CMakeFiles/cxx_static.dir/filesystem/operations.cpp.o -c /home/b/sanitizer-x86_64-linux/build/llvm-project/libcxx/src/filesystem/operations.cpp
/home/b/sanitizer-x86_64-linux/build/llvm-project/libcxx/src/filesystem/operations.cpp:246:46: error: cannot initialize a parameter of type '__off64_t *' (aka 'long long *') with an rvalue of type 'off_t *' (aka 'long *')
  246 |     if ((res = ::copy_file_range(read_fd.fd, &off_in, write_fd.fd, &off_out, count, 0)) == -1) {
      |                                              ^~~~~~~
/usr/lib/gcc-cross/i686-linux-gnu/14/../../../../i686-linux-gnu/include/unistd.h:1142:49: note: passing argument to parameter '__pinoff' here
 1142 | ssize_t copy_file_range (int __infd, __off64_t *__pinoff,
      |                                                 ^
1 error generated.
```

vitalybuka commented:
Looks like on linux this code must compile with _FILE_OFFSET_BITS 64

Jannik2099 commented:
> Looks like on linux this code must compile with _FILE_OFFSET_BITS 64

largefile handling isn't that simple on linux sadly, especially once multiple libcs are involved.

I've changed it to just use the linux-specific `loff_t`, which is also what we did for libstdc++. See the linked PR

vvereschaka commented:
We got the following build error for the cross builds (win to armv7 linux)

https://lab.llvm.org/buildbot/#/builders/38/builds/1661
```
AILED: libcxx/src/CMakeFiles/cxx_static.dir/filesystem/operations.cpp.o 
C:\buildbot\as-builder-1\x-armv7l\build\.\bin\clang++.exe --target=armv7-unknown-linux-gnueabihf -DLIBCXX_BUILDING_LIBCXXABI -DLIBC_NAMESPACE=__llvm_libc_common_utils -D_DEBUG -D_GLIBCXX_ASSERTIONS -D_LIBCPP_BUILDING_LIBRARY -D_LIBCPP_HAS_NO_PRAGMA_SYSTEM_HEADER -D_LIBCPP_LINK_PTHREAD_LIB -D_LIBCPP_LINK_RT_LIB -D_LIBCPP_REMOVE_TRANSITIVE_INCLUDES -D__STDC_CONSTANT_MACROS -D__STDC_FORMAT_MACROS -D__STDC_LIMIT_MACROS -IC:/buildbot/as-builder-1/x-armv7l/llvm-project/libcxx/src -IC:/buildbot/as-builder-1/x-armv7l/build/include/armv7-unknown-linux-gnueabihf/c++/v1 -IC:/buildbot/as-builder-1/x-armv7l/build/include/c++/v1 -IC:/buildbot/as-builder-1/x-armv7l/llvm-project/libcxxabi/include -IC:/buildbot/as-builder-1/x-armv7l/llvm-project/runtimes/cmake/Modules/../../../libc -fPIC -fno-semantic-interposition -fvisibility-inlines-hidden -Werror=date-time -Werror=unguarded-availability-new -Wall -Wextra -Wno-unused-parameter -Wwrite-strings -Wcast-qual -Wmissing-field-initializers -Wimplicit-fallthrough -Wcovered-switch-default -Wno-noexcept-type -Wnon-virtual-dtor -Wdelete-non-virtual-dtor -Wsuggest-override -Wstring-conversion -Wmisleading-indentation -Wctad-maybe-unsupported -fdiagnostics-color -ffunction-sections -fdata-sections -O3 -DNDEBUG -UNDEBUG -faligned-allocation -nostdinc++ -fvisibility-inlines-hidden -fvisibility=hidden -fsized-deallocation -Wall -Wextra -Wnewline-eof -Wshadow -Wwrite-strings -Wno-unused-parameter -Wno-long-long -Werror=return-type -Wextra-semi -Wundef -Wunused-template -Wformat-nonliteral -Wzero-length-array -Wdeprecated-redundant-constexpr-static-def -Wno-user-defined-literals -Wno-covered-switch-default -Wno-suggest-override -Wno-error -std=c++2b -MD -MT libcxx/src/CMakeFiles/cxx_static.dir/filesystem/operations.cpp.o -MF libcxx\src\CMakeFiles\cxx_static.dir\filesystem\operations.cpp.o.d -o libcxx/src/CMakeFiles/cxx_static.dir/filesystem/operations.cpp.o -c C:/buildbot/as-builder-1/x-armv7l/llvm-project/libcxx/src/filesystem/operations.cpp
C:/buildbot/as-builder-1/x-armv7l/llvm-project/libcxx/src/filesystem/operations.cpp:246:46: error: cannot initialize a parameter of type '__off64_t *' (aka 'long long *') with an rvalue of type 'off_t *' (aka 'long *')
  246 |     if ((res = ::copy_file_range(read_fd.fd, &off_in, write_fd.fd, &off_out, count, 0)) == -1) {
      |                                              ^~~~~~~
c:/buildbot/fs/jetson-tk1-arm-ubuntu/usr/include\unistd.h:1110:49: note: passing argument to parameter '__pinoff' here
 1110 | ssize_t copy_file_range (int __infd, __off64_t *__pinoff,
      |                                                 ^
1 error generated.
```

kongy commented:
`copy_file_range` was introduced in glibc 2.27. This change breaks environments with older glibc than that.

```
libcxx/src/filesystem/operations.cpp:246:18: error: no member named 'copy_file_range' in the global namespace
  246 |     if ((res = ::copy_file_range(read_fd.fd, &off_in, write_fd.fd, &off_out, count, 0)) == -1) {
      |                ~~^
1 error generated.
```

vitalybuka commented:
> We got the following build error for the cross builds (win to armv7 linux)

The is https://github.com/llvm/llvm-project/pull/121855 for that

fmayer commented:
We got another error related to this on the sanitizer buildbot

```
FAILED: libcxx/src/CMakeFiles/cxx_static.dir/filesystem/operations.cpp.o 
/home/b/sanitizer-x86_64-linux/build/build_default/bin/clang++ -DLIBCXX_BUILDING_LIBCXXABI -DLIBC_NAMESPACE=__llvm_libc_common_utils -D_LIBCPP_BUILDING_LIBRARY -D_LIBCPP_HAS_NO_PRAGMA_SYSTEM_HEADER -D_LIBCPP_LINK_PTHREAD_LIB -D_LIBCPP_LINK_RT_LIB -D_LIBCPP_REMOVE_TRANSITIVE_INCLUDES -D__STDC_CONSTANT_MACROS -D__STDC_FORMAT_MACROS -D__STDC_LIMIT_MACROS -I/home/b/sanitizer-x86_64-linux/build/llvm-project/libcxx/src -I/home/b/sanitizer-x86_64-linux/build/build_default/runtimes/runtimes-bins/compiler-rt/lib/sanitizer_common/symbolizer/RTSanitizerCommonSymbolizerInternal.i386/symbolizer/libcxx/include/c++/v1 -I/home/b/sanitizer-x86_64-linux/build/llvm-project/libcxxabi/include -I/home/b/sanitizer-x86_64-linux/build/llvm-project/runtimes/cmake/Modules/../../../libc -fvisibility-inlines-hidden -Werror=date-time -Werror=unguarded-availability-new -Wall -Wextra -Wno-unused-parameter -Wwrite-strings -Wcast-qual -Wmissing-field-initializers -Wimplicit-fallthrough -Wcovered-switch-default -Wno-noexcept-type -Wnon-virtual-dtor -Wdelete-non-virtual-dtor -Wsuggest-override -Wstring-conversion -Wmisleading-indentation -Wctad-maybe-unsupported -fdiagnostics-color -ffunction-sections -fdata-sections -m32 -U_FILE_OFFSET_BITS -fPIC -flto -Oz -g0 -DNDEBUG -target i386-unknown-linux-gnu -Wno-unused-command-line-argument -include /home/b/sanitizer-x86_64-linux/build/llvm-project/compiler-rt/lib/sanitizer_common/symbolizer/../sanitizer_redefine_builtins.h -DSANITIZER_COMMON_REDEFINE_BUILTINS_IN_STD -Wno-language-extension-token -Wno-macro-redefined -std=c++23 -faligned-allocation -nostdinc++ -fvisibility-inlines-hidden -fvisibility=hidden -fsized-deallocation -Wall -Wextra -Wnewline-eof -Wshadow -Wwrite-strings -Wno-unused-parameter -Wno-long-long -Werror=return-type -Wextra-semi -Wundef -Wunused-template -Wformat-nonliteral -Wzero-length-array -Wdeprecated-redundant-constexpr-static-def -Wno-user-defined-literals -Wno-covered-switch-default -Wno-suggest-override -Wno-error -fno-exceptions -fno-rtti -MD -MT libcxx/src/CMakeFiles/cxx_static.dir/filesystem/operations.cpp.o -MF libcxx/src/CMakeFiles/cxx_static.dir/filesystem/operations.cpp.o.d -o libcxx/src/CMakeFiles/cxx_static.dir/filesystem/operations.cpp.o -c /home/b/sanitizer-x86_64-linux/build/llvm-project/libcxx/src/filesystem/operations.cpp
/home/b/sanitizer-x86_64-linux/build/llvm-project/libcxx/src/filesystem/operations.cpp:246:46: error: cannot initialize a parameter of type '__off64_t *' (aka 'long long *') with an rvalue of type 'off_t *' (aka 'long *')
  246 |     if ((res = ::copy_file_range(read_fd.fd, &off_in, write_fd.fd, &off_out, count, 0)) == -1) {
      |                                              ^~~~~~~
/usr/lib/gcc-cross/i686-linux-gnu/14/../../../../i686-linux-gnu/include/unistd.h:1142:49: note: passing argument to parameter '__pinoff' here
 1142 | ssize_t copy_file_range (int __infd, __off64_t *__pinoff,
      |                                                 ^
1 error generated.
```

Jannik2099 commented:
> We got another error related to this on the sanitizer buildbot (https://lab.llvm.org/buildbot/#/builders/66/builds/8314)

sorry, this is also fixed in the pending #121855

ldionne commented:
I merged https://github.com/llvm/llvm-project/pull/121855, thanks @Jannik2099 for the quick fix!
